src/mds/schedule/scanScheduler.cpp

```diff
@@ -21,27 +21,167 @@
  */
 
 #include <glog/logging.h>
+#include <algorithm>
+#include "src/common/timeutility.h"
 #include "src/mds/schedule/scheduler.h"
+#include "src/mds/schedule/operatorFactory.h"
 
 namespace curve {
 namespace mds {
 namespace schedule {
+
 int ScanScheduler::Schedule() {
-    return 0;
+    LOG(INFO) << "ScanScheduelr begin.";
+
+    auto currentHour = ::curve::common::TimeUtility::GetCurrentHour();
+    bool duringScanTime = currentHour >= scanStartHour_ &&
+                          currentHour <= scanEndHour_;
+
+    auto count = 0;
+    ::curve::mds::topology::LogicalPool lpool;
+    auto logicPoolIds = topo_->GetLogicalpools();
+    for (const auto& lpid : logicPoolIds) {
+        CopySetInfos copysets2start, copysets2cancel;
+        auto copysetInfos = topo_->GetCopySetInfosInLogicalPool(lpid);
+        topo_->GetLogicalPool(lpid, &lpool);
+        if (!duringScanTime || !lpool.ScanEnable()) {
+            for (const auto& copysetInfo : copysetInfos) {
+                if (copysetInfo.scaning || ReadyToScan(copysetInfo)) {
+                    copysets2cancel.push_back(copysetInfo);
+                }
+            }
+        } else {
+            SelectCopysetsForScan(
+                copysetInfos, &copysets2start, &copysets2cancel);
+        }
+
+        count += GenScanOperator(copysets2start,
+                                 ConfigChangeType::START_SCAN_PEER);
+        count += GenScanOperator(copysets2cancel,
+                                 ConfigChangeType::CANCEL_SCAN_PEER);
+    }
+
+    LOG(INFO) << "ScanScheduelr generate " << count << " at this round";
+    return 1;
 }
 
-int64_t ScanScheduler::GetRunningInterval() {
-    return runInterval_;
+bool ScanScheduler::ReadyToScan(CopySetInfo copysetInfo) {
+    Operator op;
+    if (opController_->GetOperatorById(copysetInfo.id, &op)) {
+        auto step = dynamic_cast<ScanPeer*>(op.step.get());
+        return nullptr != step && step->IsStartScan();
+    } else if (copysetInfo.HasCandidate()) {
+        return copysetInfo.configChangeInfo.type() ==
+               ConfigChangeType::START_SCAN_PEER;
+    }
+
+    return false;
+}
+
+void ScanScheduler::SelectCopysetsToStartScan(CopySetInfos copysetInfos,
```

Comment:
`copysetInfos` and `selected` are passing by value, they can be very large, so when calling this function, `copysetInfos` and `selected` will construct from its arguments, it will cause a lot of memory copy.

Reply:
> `copysetInfos` and `selected` are passing by value, they can be very large, so when calling this function, `copysetInfos` and `selected` will construct from its arguments, it will cause a lot of memory copy.

yeap, i has improve it.